File: src/Aiko.Core/ClaudeInstall.cs
namespace Aiko.Core;

/// Where Claude Code is, and where a new environment's folder goes.
public static class ClaudeInstall
{
    /// The PATH a terminal opened right now would get: the machine part, then the user part, with
    /// variables expanded. Aiko itself may have started before Claude Code was installed, and its
    /// own PATH is then out of date: the wizard would wait forever for a Claude Code that is there.
    public static string FreshPath(
        PlatformConventions platform, string? machinePath, string? userPath, Func<string, string> expand) =>
        string.Join(platform.PathListSeparator, new[] { machinePath, userPath }
            .Where(part => !string.IsNullOrWhiteSpace(part))
            .Select(part => expand(part!)));

    /// The real Claude Code on that PATH, never Aiko's own shim, or where the native installer puts
    /// it when PATH has not caught up.
    public static string? Find(
        PlatformConventions platform, string freshPath, string commandFolder, string userProfile, Func<string, bool> exists)
    {
        var onPath = RealClaude.Find(platform, freshPath, [commandFolder], exists);
        if (onPath is not null)
        {
            return onPath;
        }

        var native = platform.Join(userProfile, ".local", "bin", RealClaude.ExecutableName(platform));
        return exists(native) ? native : null;
    }

    /// .claude-work for "Work", .claude-osnovnaya for "Основная". A name already taken gets a
    /// number, so a new environment never lands in somebody's existing account folder.
    public static string NewConfigFolder(
        PlatformConventions platform, string environmentName, string userProfile, Func<string, bool> folderExists)
    {
        var slug = LaunchCommand.Slug(environmentName);
        var stem = platform.Join(userProfile, ClaudeConfigFolder.NamedFolderName(slug.Length > 0 ? slug : "env"));

        var candidate = stem;
        for (var n = 2; folderExists(candidate); n++)
        {
            candidate = stem + "-" + n;
        }

        return candidate;
    }

    public const string CredentialsFileName = ".credentials.json";

    /// Signed in means Claude Code wrote its credentials file. Only that the file exists is looked
    /// at; it is never opened here.
    public static string CredentialsPathIn(PlatformConventions platform, string configFolder) =>
        platform.Join(configFolder, CredentialsFileName);
}

File: mac/Sources/AikoKit/ClaudeInstall.swift
import Foundation

/// Where Claude Code is, and where a new environment's folder goes.
public enum ClaudeInstall {
    /// The PATH a terminal opened right now would get: the machine part, then the user part, with
    /// variables expanded. Aiko itself may have started before Claude Code was installed, and its
    /// own PATH is then out of date: the wizard would wait forever for a Claude Code that is there.
    public static func freshPath(
        _ platform: PlatformConventions,
        machinePath: String?,
        userPath: String?,
        expand: (String) -> String
    ) -> String {
        [machinePath, userPath]
            .compactMap { $0 }
            .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .map(expand)
            .joined(separator: String(platform.pathListSeparator))
    }

    /// The real Claude Code on that PATH, never Aiko's own shim, or where the native installer puts
    /// it when PATH has not caught up.
    public static func find(
        _ platform: PlatformConventions,
        freshPath: String,
        commandFolder: String,
        userProfile: String,
        exists: (String) -> Bool
    ) -> String? {
        if let onPath = RealClaude.find(platform, freshPath, [commandFolder], exists) {
            return onPath
        }

        let native = platform.join(userProfile, ".local", "bin", RealClaude.executableName(platform))
        return exists(native) ? native : nil
    }

    /// .claude-work for "Work", .claude-osnovnaya for "Основная". A name already taken gets a
    /// number, so a new environment never lands in somebody's existing account folder.
    public static func newConfigFolder(
        _ platform: PlatformConventions,
        environmentName: String,
        userProfile: String,
        folderExists: (String) -> Bool
    ) -> String {
        let slug = LaunchCommand.slug(environmentName)
        let stem = platform.join(userProfile, ClaudeConfigFolder.namedFolderName(slug.isEmpty ? "env" : slug))

        var candidate = stem
        var n = 2
        while folderExists(candidate) {
            candidate = stem + "-" + String(n)
            n += 1
        }

        return candidate
    }

    public static let credentialsFileName = ".credentials.json"

    /// Signed in means Claude Code wrote its credentials file. Only that the file exists is looked
    /// at; it is never opened here.
    public static func credentialsPathIn(_ platform: PlatformConventions, _ configFolder: String) -> String {
        platform.join(configFolder, credentialsFileName)
    }
}
